oxide-auth/src/primitives/registrar.rs

```diff
@@ -291,107 +286,69 @@ pub trait PasswordPolicy: Send + Sync {
     fn check(&self, client_id: &str, passphrase: &[u8], stored: &[u8]) -> Result<(), RegistrarError>;
 }
 
-/// Store passwords using `Pbkdf2` to derive the stored value.
-///
-/// Each instantiation generates a 16 byte random salt and prepends this additionally with the
-/// username. This combined string is then used as the salt using the passphrase as the secret to
-/// derive the output. The iteration count defaults to `65536` but can be customized.
-pub struct Pbkdf2 {
-    /// A prebuilt random, or constructing one as needed.
-    random: Option<SystemRandom>,
-    iterations: NonZeroU32,
-}
-
-impl Default for Pbkdf2 {
-    fn default() -> Self {
-        Pbkdf2 {
-            random: Some(SystemRandom::new()),
-            .. *PBKDF2_DEFAULTS
-        }
-    }
+/// Store passwords using `Argon2` to derive the stored value.
+#[derive(Debug)]
+pub struct Argon2<'a> {
+    secret_key: SecretKey<'a>
 }
 
-impl Clone for Pbkdf2 {
-    fn clone(&self) -> Self {
-        Pbkdf2 {
-            random: Some(SystemRandom::new()),
-            .. *self
+impl<'a> Default for Argon2<'a> {
+    fn default() -> Self {   
+        Argon2 {
+            secret_key: "Please provide a SecretKey! This is for your own protection.".into()
         }
     }
 }
 
-impl fmt::Debug for Pbkdf2 {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("Pbkdf2")
-            .field("iterations", &self.iterations)
-            .field("random", &())
-            .finish()
+impl From<std::string::FromUtf8Error> for RegistrarError {
+    fn from(_err: std::string::FromUtf8Error) -> Self {
+        RegistrarError::PrimitiveError
     }
 }
 
-impl Pbkdf2 {
-    /// Set the iteration count to `(1 << strength)`.
-    ///
-    /// This function will panic when the `strength` is larger or equal to `32`.
-    pub fn set_relative_strength(&mut self, strength: u8) {
-        assert!(strength < 32, "Strength value out of range (0-31): {}", strength);
-        self.iterations = NonZeroU32::new(1u32 << strength).unwrap();
-    }
-
-    fn salt(&self, user_identifier: &[u8]) -> Vec<u8> {
-        let mut vec = Vec::with_capacity(user_identifier.len() + 64);
-        let mut rnd_salt = [0; 16];
-
-        match self.random.as_ref() {
-            Some(random) => random.fill(&mut rnd_salt),
-            None => SystemRandom::new().fill(&mut rnd_salt),
-        }.expect("Failed to property initialize password storage salt");
-
-        vec.extend_from_slice(user_identifier);
-        vec.extend_from_slice(&rnd_salt[..]);
-        vec
+impl From<argonautica::Error> for RegistrarError {
```

Comment:
This impl leaks the implementation detail of `argonautica` into the public interface. That should be avoided.

Reply:
Removed